CSI block volume refactor to fix target path

mkimuram opened the pull request:
**What this PR does / why we need it**:
Fix for adding block volume support to CSI RBD driver

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #68424 

**Special notes for your reviewer**:
/sig storage

This PR fixes following 3 issues:
[Issues]
1. ```symlinkPath``` used in ```MapDevice``` doesn't match to the one ```makeBlockVolumes``` expects
2. ~~No need to create file in neither ```targetBlockFilePath``` nor ```globalMapPathBlockFile```~~  It is not an issue. CSI spec requires CO to create an empty file to bind mount the device on, so the file should be created.
3. ```EvalHostSymlinksdevice``` fails for csi drivers that don't implement ```NodeStageVolume```

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

mkimuram commented:
@wongma7 @vladimirvivien

neolit123 commented:
/kind bug

vladimirvivien commented on pkg/volume/csi/csi_block.go:
```diff
@@ -56,12 +56,12 @@ func (m *csiBlockMapper) GetGlobalMapPath(spec *volume.Spec) (string, error) {
-	specName := m.specName
-	glog.V(4).Infof(log("blockMapper.GetPodDeviceMapPath [path=%s; name=%s]", path, specName))
-	return path, specName
+	fileName := "file"
```
Returning the value `"file"` seems extremely specific and is may not be useful for all drivers .  Returning the volumeID may be needed during operations such as tearDown.

vladimirvivien commented on pkg/volume/csi/csi_block.go:
```diff
@@ -139,18 +139,6 @@ func (m *csiBlockMapper) SetUpDevice() (string, error) {
 	}
 	glog.V(4).Info(log("blockMapper.SetupDevice created global device map path successfully [%s]", globalMapPath))
 
-	// create block device file
```
Delegating the creation of block device to the Driver may be a good approach.  However, before this is removed, I think it should be cleared/documented that this step will be delegated to the driver.

mkimuram replied:
I will add comments to ```SetUpDevice``` and ```MapDevice``` before calling ```NodeStageVolume``` and ```NodePublishVolume```.

vladimirvivien commented on pkg/volume/csi/csi_block.go:
```diff
@@ -201,8 +189,8 @@ func (m *csiBlockMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, vol
 	globalMapPathBlockFile := devicePath
-	dir, _ := m.GetPodDeviceMapPath()
-	targetBlockFilePath := filepath.Join(dir, "file")
+	dir, file := m.GetPodDeviceMapPath()
```
Again returning `"file"` will not work properly for drivers that needs volumeID during Teardown.

mkimuram replied:
I will rename ```file``` to ```volumeID``` when ```GetPodDeviceMapPath``` is changed to return ```volumeID```.

vladimirvivien reviewed (commented):
Overall I think this is gettin closer.  Let me think about the issue of delegating the creation of the block file to the driver.  While this works, it means that driver pods will be forced to have privileged access to the node.

vladimirvivien commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -887,24 +888,26 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
 
-		// When kubelet is containerized, devicePath may be a symlink at a place unavailable to
-		// kubelet, so evaluate it on the host and expect that it links to a device in /dev,
+		// Map device to global and pod device map path
```
Changes here will affect how all block volume devices work including in-tree drivers.  Any changes here should be tested to make sure other drivers are working as before.

mkimuram replied:
As for in-tree drivers, I added [e2e tests for block volume](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/in_tree_volumes.go#L51). So, it would be tested there. (Block volume feature was missed to be beta in v1.12, so it won't be automatically tested by CSI, though.)
In addition, it won't be tested with combinations of containerized kubelet, so we might need to find a way to test with containerized kubelet test environment. Let me think how to test it.

As for csi drivers, I'm working on to apply the same testsuites that in-tree drivers do to CSI drivers, [here](https://github.com/kubernetes/kubernetes/issues/68024) and it will make all csi drivers to be tested for block volume. (Actually, no block volume supported CSI drivers there, though.)

vladimirvivien reviewed (commented):
This portion needs careful attention not to effect other drivers.

mkimuram committed:
Fix for adding block volume support to CSI RBD driver

mkimuram commented:
@vladimirvivien 

> Returning the volumeID may be needed during operations such as tearDown.

I looked through the ```Teardown``` logic for block volume again as you advised me. As a result, I came to the conclusion that it would be better to avoid delegating the creation of symlink to the path provided by ```GetPodDeviceMapPath``` and ```GetGlobalMapPath``` to each CSI driver. 
Instead, CSI driver should create symlink or bind mount their actual device to another path, like ```stagingPath``` and ```publishPath``` in ```SetUpDevice``` through ```NodeStageVolume``` and ```NodePublishVolume```, then ```csiBlockMapper``` should create symlink to the path provided by ```GetPodDeviceMapPath``` and ```GetGlobalMapPath``` in ```MapDevice```, as existing in-tree drivers do.
 By doing so, we don't need to touch much on existing block volume code. 

I implemented above way in [another commit in a separate branch](https://github.com/mkimuram/kubernetes/commit/1e591792edd41726c26b5355a1888becaa746136). PTAL

Summary of the difference of the concept is as below:
(I hope that it will help you review the code.)

- Map:
```
  mapVolumeFunc
    SetUpDevice
      NodeStageVolume   (Attach device to node and create symlink to the path provided by GetGlobalMapPath)
    MapDevice
      NodePublishVolume (Create symlink to the path provided by GetPodDeviceMapPath)
```

- Unmap:
```
  unmapVolumeFunc
    og.blkUtil.UnmapDevice (Delete symlink from the path provided by GetPodDeviceMapPath)
    og.blkUtil.UnmapDevice (Delete symlink from the path provided by GetGlobalMapPath)
  unmapDeviceFunc
    TearDownDevice
      NodeUnstageVolume
      NodeUnpublishVolume  (Detach device from node)
```


[New Implementation above]
- Map:
```
  mapVolumeFunc
    SetUpDevice
      NodeStageVolume       (Attach device to node and Create symlink to the stagingPath)
      NodePublishVolume     (Create symlink to the path publishPath)
    MapDevice
      ioutil.MapBlockVolume (Create symlink to the path provided by GetGlobalMapPath and GetPodDeviceMapPath)
```

- Unmap:
```
  unmapVolumeFunc
    og.blkUtil.UnmapDevice (Delete symlink to the path provided by GetPodDeviceMapPath)
    og.blkUtil.UnmapDevice (Delete symlink to the path provided by GetGlobalMapPath)
  unmapDeviceFunc
    TearDownDevice
      NodeUnpublishVolume  (Delete symlink from stagingPath)
      NodeUnstageVolume    (Delete symlink from publishPath and detach device from node)
```

vladimirvivien commented:
@mkimuram I am reviewing your proposed changes today. I will post feedback here and other PR where appropriate.

vladimirvivien commented:
@mkimuram 
Ok overall the second approach looks clean and workable.  The code organization looks great.  I do have couple of things I want to review

- The following changes from `volumeToDetach.VolumeName` -> `volumeToDetach.VolumeSpec.Name()` If these are not equivalent and equal, it should not be changed. https://github.com/mkimuram/kubernetes/commit/1e591792edd41726c26b5355a1888becaa746136#diff-450e811a4953f760ff1594ede8b2037eR1063

- Retrieve the Attachment object in `SetupDevice` and pass attachment to methods `stageVolumeForBloc` and `publishVolumeForBlock`.  That way the attachment lookup does not occur twice (once in each method call). https://github.com/mkimuram/kubernetes/blob/1e591792edd41726c26b5355a1888becaa746136/pkg/volume/csi/csi_block.go#L272

Other than that, I think this maybe workable.

Good progress.

mkimuram commented:
@vladimirvivien 

Thank your for your comment.

>Ok overall the second approach looks clean and workable. The code organization looks great. I do have couple of things I want to review

I will fix based on your comment and make PR.

> The following changes from volumeToDetach.VolumeName -> volumeToDetach.VolumeSpec.Name() If these are not equivalent and equal, it should not be changed. mkimuram@1e59179#diff-450e811a4953f760ff1594ede8b2037eR1063

Actually, as for the changes in operation_generator.go, I found another two issues below while testing with my rbd prototype codes and they are the fixes for the issues.

1. The first argument for ```NewBlockVolumeUnmapper``` is inconsistent between ```GenerateUnmapVolumeFunc``` and ```GenerateUnmapDeviceFunc```, as a result, ```loadVolumeData``` in ```NewBlockVolumeUnmapper``` fails.

--> As specName seems to be expected there, I fixed as so. 

2. ```TearDownDevice``` fails due to device in use and it is caused by descriptor lock

--> As descriptor lock seems required to be released before ```TearDownDevice```, I fixed as so. 

Anyway, we need to check and test them carefully, whether they won't affect any existing drivers.
(Without them, the changes were only in csi_block.go, we would be easier to decide applying changes.)

> Retrieve the Attachment object in SetupDevice and pass attachment to methods stageVolumeForBloc and publishVolumeForBlock. That way the attachment lookup does not occur twice (once in each method call). https://github.com/mkimuram/kubernetes/blob/1e591792edd41726c26b5355a1888becaa746136/pkg/volume/csi/csi_block.go#L272

I will fix as so. Thank you for pointing it out.

vladimirvivien commented:
/ok-to-test

mkimuram commented:
@vladimirvivien 

I've updated the commit to the second version and add unit test for it. PTAL

Please also confirm the log below.

<details>

1. Console log
```
# kubectl create -f sample/pod-block.yaml
# kubectl describe pod csirbd-block-demo-pod

<snip>

Events:                                                                                                         
  Type    Reason                  Age   From                     Message                                        
  ----    ------                  ----  ----                     -------                                        
  Normal  Scheduled               25s   default-scheduler        Successfully assigned default/csirbd-block-demo
-pod to 127.0.0.1                                                                                               
  Normal  SuccessfulAttachVolume  25s   attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-8
db97c58-c00a-11e8-b6ef-525400b854f0"                                                                            
  Normal  SuccessfulMountVolume   9s    kubelet, 127.0.0.1       MapVolume.MapDevice succeeded for volume "pvc-8
db97c58-c00a-11e8-b6ef-525400b854f0" globalMapPath "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc
-8db97c58-c00a-11e8-b6ef-525400b854f0/dev"                                                                      
  Normal  SuccessfulMountVolume   9s    kubelet, 127.0.0.1       MapVolume.MapDevice succeeded for volume "pvc-8
db97c58-c00a-11e8-b6ef-525400b854f0" volumeMapPath "/var/lib/kubelet/pods/91a17d03-c00a-11e8-b6ef-525400b854f0/v
olumeDevices/kubernetes.io~csi"                                                                                 
  Normal  Pulling                 7s    kubelet, 127.0.0.1       pulling image "nginx"
  Normal  Pulled                  4s    kubelet, 127.0.0.1       Successfully pulled image "nginx"
  Normal  Created                 4s    kubelet, 127.0.0.1       Created container
  Normal  Started                 4s    kubelet, 127.0.0.1       Started container
# kubectl delete -f sample/pod-block.yaml
```
2. Check if the mapped device can be accessible inside pod
```
# kubectl exec -it csirbd-block-demo-pod bash 
root@csirbd-block-demo-pod:/# test -b /mnt/block
root@csirbd-block-demo-pod:/# echo $?
0
root@csirbd-block-demo-pod:/# test -d /mnt/block                                                                
root@csirbd-block-demo-pod:/# echo $?                                                                           
1
# exit
exit
```
3. Path (Before pod deletion)
  3.1 symlink to the path provided by GetGlobalMapPath 
```
# ls -lR /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b
6ef-525400b854f0/                                                                                               
/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/:
total 8
drwxr-x---. 2 root root 4096 Sep 24 15:00 data
drwxr-x---. 2 root root 4096 Sep 24 15:00 dev

/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/data:
total 4
-rw-r--r--. 1 root root 257 Sep 24 15:00 vol_data.json

/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/dev:
total 0
lrwxrwxrwx. 1 root root 9 Sep 24 15:00 91a17d03-c00a-11e8-b6ef-525400b854f0 -> /dev/rbd0
```
  3.2 symlink to the path provided by ```getPublishPath```
```
# ls -lR /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/                
/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/:
total 0
lrwxrwxrwx. 1 root root 9 Sep 24 15:00 pvc-8db97c58-c00a-11e8-b6ef-525400b854f0 -> /dev/rbd0
```
  3.3 symlink to the path provided by ```GetPodDeviceMapPath```
```
# ls -lR /var/lib/kubelet/pods/91a17d03-c00a-11e8-b6ef-525400b854f0/volumeDevices/kubernet
es.io~csi/                                                                                                      
/var/lib/kubelet/pods/91a17d03-c00a-11e8-b6ef-525400b854f0/volumeDevices/kubernetes.io~csi/:
total 0
lrwxrwxrwx. 1 root root 9 Sep 24 15:00 pvc-8db97c58-c00a-11e8-b6ef-525400b854f0 -> /dev/rbd0
```
4. Path (After pod deletion)
  4.1 symlink to the path provided by GetGlobalMapPath 
```
# ls -lR /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/
/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/:
total 4
drwxr-x---. 2 root root 4096 Sep 24 15:00 data

/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/data:
total 4
-rw-r--r--. 1 root root 257 Sep 24 15:00 vol_data.json
```
  4.2 symlink to the path provided by ```getPublishPath```
```
# ls -lR /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/  
/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/:
total 0
```
  4.3 symlink to the path provided by ```GetPodDeviceMapPath```
```
# ls -lR /var/lib/kubelet/pods/91a17d03-c00a-11e8-b6ef-525400b854f0/
ls: cannot access '/var/lib/kubelet/pods/91a17d03-c00a-11e8-b6ef-525400b854f0/': No such file or directory
```
5. Log:
  5.1 Map:
```
# less /tmp/kubelet.log 

<snip>

I0924 15:00:41.926647  196280 reconciler.go:207] operationExecutor.VerifyControllerAttachedVolume started for volume "pvc-8db97c58-c00a-11e8-b6ef-52540
0b854f0" (UniqueName: "kubernetes.io/csi/csi-rbdplugin^csi-rbd-8dd52e4a-c00a-11e8-b3eb-525400b854f0") pod "csirbd-block-demo-pod" (UID: "91a17d03-c00a-
11e8-b6ef-525400b854f0") 
I0924 15:00:41.931329  196280 operation_generator.go:1206] Controller attach succeeded for volume "pvc-8db97c58-c00a-11e8-b6ef-525400b854f0" (UniqueNam
e: "kubernetes.io/csi/csi-rbdplugin^csi-rbd-8dd52e4a-c00a-11e8-b3eb-525400b854f0") pod "csirbd-block-demo-pod" (UID: "91a17d03-c00a-11e8-b6ef-525400b85
4f0") device path: "csi-665e3df0874c62c78bffd36eea417afd2731d4af771ad7ad6a1fe364f5c6a119"
I0924 15:00:42.027941  196280 reconciler.go:252] operationExecutor.MountVolume started for volume "pvc-8db97c58-c00a-11e8-b6ef-525400b854f0" (UniqueNam
e: "kubernetes.io/csi/csi-rbdplugin^csi-rbd-8dd52e4a-c00a-11e8-b3eb-525400b854f0") pod "csirbd-block-demo-pod" (UID: "91a17d03-c00a-11e8-b6ef-525400b85
4f0") 
I0924 15:00:42.028176  196280 operation_generator.go:862] MapVolume.WaitForAttach entering for volume "pvc-8db97c58-c00a-11e8-b6ef-525400b854f0" (Uniqu
eName: "kubernetes.io/csi/csi-rbdplugin^csi-rbd-8dd52e4a-c00a-11e8-b3eb-525400b854f0") pod "csirbd-block-demo-pod" (UID: "91a17d03-c00a-11e8-b6ef-52540
0b854f0") DevicePath "csi-665e3df0874c62c78bffd36eea417afd2731d4af771ad7ad6a1fe364f5c6a119"
I0924 15:00:42.032745  196280 operation_generator.go:871] MapVolume.WaitForAttach succeeded for volume "pvc-8db97c58-c00a-11e8-b6ef-525400b854f0" (Uniq
ueName: "kubernetes.io/csi/csi-rbdplugin^csi-rbd-8dd52e4a-c00a-11e8-b3eb-525400b854f0") pod "csirbd-block-demo-pod" (UID: "91a17d03-c00a-11e8-b6ef-5254
00b854f0") DevicePath "csi-665e3df0874c62c78bffd36eea417afd2731d4af771ad7ad6a1fe364f5c6a119"
I0924 15:00:42.037884  196280 csi_block.go:103] kubernetes.io/csi: blockMapper.stageVolumeForBlock STAGE_UNSTAGE_VOLUME capability not set. Skipping Mo
untDevice...
I0924 15:00:42.451335  196280 prober.go:118] Liveness probe for "kube-dns-596fbb8fbd-pxzct_kube-system(4fe395a1-c00a-11e8-b6ef-525400b854f0):sidecar" s
ucceeded
I0924 15:00:42.576578  196280 operation_generator.go:935] MapVolume.MapDevice succeeded for volume "pvc-8db97c58-c00a-11e8-b6ef-525400b854f0" (UniqueNa
me: "kubernetes.io/csi/csi-rbdplugin^csi-rbd-8dd52e4a-c00a-11e8-b3eb-525400b854f0") pod "csirbd-block-demo-pod" (UID: "91a17d03-c00a-11e8-b6ef-525400b8
54f0") volumeMapPath "/var/lib/kubelet/pods/91a17d03-c00a-11e8-b6ef-525400b854f0/volumeDevices/kubernetes.io~csi"
I0924 15:00:42.576694  196280 server.go:460] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"csirbd-block-demo-pod", UID:"91a17d03-c00a
-11e8-b6ef-525400b854f0", APIVersion:"v1", ResourceVersion:"440", FieldPath:""}): type: 'Normal' reason: 'SuccessfulMountVolume' MapVolume.MapDevice su
cceeded for volume "pvc-8db97c58-c00a-11e8-b6ef-525400b854f0" volumeMapPath "/var/lib/kubelet/pods/91a17d03-c00a-11e8-b6ef-525400b854f0/volumeDevices/k
ubernetes.io~csi"
I0924 15:00:42.576752  196280 server.go:460] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"csirbd-block-demo-pod", UID:"91a17d03-c00a
-11e8-b6ef-525400b854f0", APIVersion:"v1", ResourceVersion:"440", FieldPath:""}): type: 'Normal' reason: 'SuccessfulMountVolume' MapVolume.MapDevice su
cceeded for volume "pvc-8db97c58-c00a-11e8-b6ef-525400b854f0" globalMapPath "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a
-11e8-b6ef-525400b854f0/dev"
I0924 15:00:42.759437  196280 volume_manager.go:383] All volumes are attached and mounted for pod "csirbd-block-demo-pod_default(91a17d03-c00a-11e8-b6e
f-525400b854f0)"
```
  5.2 Unmap:
```
# less /tmp/kubelet.log 

<snip>

I0924 15:05:13.753726  196280 reconciler.go:181] operationExecutor.UnmountVolume started for volume "mypvc" (UniqueName: "kubernetes.io/csi/csi-rbdplug
in^csi-rbd-8dd52e4a-c00a-11e8-b3eb-525400b854f0") pod "91a17d03-c00a-11e8-b6ef-525400b854f0" (UID: "91a17d03-c00a-11e8-b6ef-525400b854f0") 
I0924 15:05:13.754407  196280 operation_generator.go:1010] UnmapVolume succeeded for volume "kubernetes.io/csi/csi-rbdplugin^csi-rbd-8dd52e4a-c00a-11e8-b3eb-525400b854f0" (OuterVolumeSpecName: "mypvc") pod "91a17d03-c00a-11e8-b6ef-525400b854f0" (UID: "91a17d03-c00a-11e8-b6ef-525400b854f0"). InnerVolumeSpecName "pvc-8db97c58-c00a-11e8-b6ef-525400b854f0". PluginName "kubernetes.io/csi", VolumeGidValue ""
E0924 15:05:13.754553  196280 plugin_watcher.go:120] error could not find plugin for deleted file /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/dev/91a17d03-c00a-11e8-b6ef-525400b854f0 when handling delete event: "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/dev/91a17d03-c00a-11e8-b6ef-525400b854f0": REMOVE
I0924 15:05:13.768872  196280 operation_generator.go:698] UnmountVolume.TearDown succeeded for volume "kubernetes.io/secret/91a17d03-c00a-11e8-b6ef-525400b854f0-default-token-ltpvh" (OuterVolumeSpecName: "default-token-ltpvh") pod "91a17d03-c00a-11e8-b6ef-525400b854f0" (UID: "91a17d03-c00a-11e8-b6ef-525400b854f0"). InnerVolumeSpecName "default-token-ltpvh". PluginName "kubernetes.io/secret", VolumeGidValue ""
I0924 15:05:13.854151  196280 reconciler.go:301] Volume detached for volume "default-token-ltpvh" (UniqueName: "kubernetes.io/secret/91a17d03-c00a-11e8-b6ef-525400b854f0-default-token-ltpvh") on node "127.0.0.1" DevicePath ""
I0924 15:05:13.854425  196280 reconciler.go:294] operationExecutor.UnmountDevice started for volume "pvc-8db97c58-c00a-11e8-b6ef-525400b854f0" (UniqueName: "kubernetes.io/csi/csi-rbdplugin^csi-rbd-8dd52e4a-c00a-11e8-b3eb-525400b854f0") on node "127.0.0.1" 
E0924 15:05:13.893282  196280 plugin_watcher.go:120] error could not find plugin for deleted file /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0 when handling delete event: "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0": REMOVE
E0924 15:05:14.117642  196280 plugin_watcher.go:120] error could not find plugin for deleted file /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0 when handling delete event: "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0": REMOVE
E0924 15:05:14.117806  196280 plugin_watcher.go:120] error could not find plugin for deleted file /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0_deleting when handling delete event: "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0_deleting": REMOVE
I0924 15:05:14.118470  196280 operation_generator.go:1435] The path isn't device path or doesn't exist. Skip checking device path: /dev/rbd0
I0924 15:05:14.118582  196280 operation_generator.go:1132] UnmapDevice succeeded for volume "pvc-8db97c58-c00a-11e8-b6ef-525400b854f0" (UniqueName: "kubernetes.io/csi/csi-rbdplugin^csi-rbd-8dd52e4a-c00a-11e8-b3eb-525400b854f0") on node "127.0.0.1" 
E0924 15:05:14.118793  196280 plugin_watcher.go:120] error could not find plugin for deleted file /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/dev when handling delete event: "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/dev": REMOVE
E0924 15:05:14.118874  196280 plugin_watcher.go:120] error could not find plugin for deleted file /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/dev when handling delete event: "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-8db97c58-c00a-11e8-b6ef-525400b854f0/dev": REMOVE
I0924 15:05:14.155345  196280 reconciler.go:301] Volume detached for volume "pvc-8db97c58-c00a-11e8-b6ef-525400b854f0" (UniqueName: "kubernetes.io/csi/csi-rbdplugin^csi-rbd-8dd52e4a-c00a-11e8-b3eb-525400b854f0") on node "127.0.0.1" DevicePath ""
```

mkimuram commented:
/retest

mkimuram commented:
@vladimirvivien

As for the change in operation_generator.go, I think that below two points are required to be confirmed:
- (a) Whether first arugument for ```NewBlockVolumeUnmapper``` in ```GenerateUnmapDeviceFunc``` can be changed safely,
- (b) Whether descriptor lock can be released safely before calling ```TearDownDevice``` and ```RemoveMapPath```

As a result of my investigation, it should be safe for both (a) and (b). (Also, I tested e2e block test for iscsi and rbd and it passed.) Let me explain the reason.

There are two types of plugins:
- (1) plugins that do nothing in ```TearDownDevice```
  - cinder
  - local
  - gce
  - aws
  - azure
- (2) plugins that do something in ```TearDownDevice```
  - iscsi
  - rbd
  - fc


And ```TearDownDevice``` is only called from ```GenerateUnmapDeviceFunc``` in operation_generator.go.
Also, ```TearDownDevice``` is the only unmapper method that is called in ```GenerateUnmapDeviceFunc```.

Therefore:
- For (1):
  - The change in how to create unmapper in ```GenerateUnmapDeviceFunc``` won't affect anything,  because they do nothing by using the unmapper created there.  ... (a) is OK for (1)
  - The change of the order of deleting descriptor lock and ```TearDownDevice``` call won't affect because they do nothing in ```TearDownDevice```. ... (b) is OK for (1)
- For (2):
  - All plugins don't use volumeName that is directly passed from unmapper, instead they use mapPath to decide which device to detach. ... (a) is OK for (2)
    [iscsi](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/iscsi/iscsi_util.go#L635)
    [rbd](https://github.com/kubernetes/kubernetes/blob/994ceaae0e70f13821e1ed3d941f089c84f2ca41/pkg/volume/rbd/rbd.go#L958)
    [fc](https://github.com/kubernetes/kubernetes/blob/994ceaae0e70f13821e1ed3d941f089c84f2ca41/pkg/volume/fc/fc_util.go#L298)
  - All plugins release descriptor lock inside ```TearDownDevice```, therefore it is safe to release it outside, 
    or in ```GenerateUnmapDeviceFunc```.  .. (b) is OK for (2)
    [iscsi](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/iscsi/iscsi_util.go#L690)
    rbd: [TearDownDevice](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/rbd/rbd.go#L962),[DetachBlockDisk](https://github.com/kubernetes/kubernetes/blob/994ceaae0e70f13821e1ed3d941f089c84f2ca41/pkg/volume/rbd/rbd_util.go#L507)
    fc: [TearDownDevice](https://github.com/kubernetes/kubernetes/blob/994ceaae0e70f13821e1ed3d941f089c84f2ca41/pkg/volume/fc/fc.go#L458), [DetachBlockFCDisk](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/fc/fc_util.go#L351)

 
For (b), we might be able to keep the descriptor lock as is in ```GenerateUnmapDeviceFunc``` and do the same implementation to (2) in csi_block.go, however, I couldn't find a good reason to do it, expect keeping operation_generator.go code as much as as is. If it would be better to keep it as is, I will change the code for (b) as so.
How do you think about it?

rootfs commented:
sorry late to the review, I haven't tested it yet but I believe it makes sense to release the loopback device before detaching the underlying device.

I see this PR does two things: a csi block refactoring and the operator generator reordering. What if breaking them up in separate PRs?

mkimuram commented on pkg/volume/csi/csi_block.go:
```diff
@@ -279,46 +200,121 @@ func (m *csiBlockMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, vol
+	// Call NodePublishVolume
+	publishPath, err := m.publishVolumeForBlock(csiSource, attachment, stagingPath)
+	if err != nil {
+		return "", err
```
@rootfs

Thank you for your comment.

IIUC, reconciler will keep calling ```MountVolume``` to try to make ASW and DSW the same, even when ```MountVolume``` fails in ```SetUpDevice```. So, even if we call ```NodeUnstageVolume``` here to unstage the volume that failed in publish, another attempts to ```MountVolume``` would stage it again. (Then, ```MountVolume``` may fail in publish again.) In addition, if ```NodeUnstageVolume``` for Nth ```MountVolume``` are called before ```NodePublishVolume``` and after ```NodeStageVolume``` for N+1th ```MountVolume```, it will make ```NodePublishVolume``` call for unstaged volume, which violates CSI spec. Therefore, just calling ```NodeUnstageVolume``` here won't role back the status properly.

However, this raises me a question whether it is safe to keep the volume staged, after the failure in publish. At a glance, it seems safe from data corruption viewpoint, for we won't directly touch to the staged device and staging path.
The worst case senarios that I noticed are below:
  - If the pvc deletion is done before the publish-failed pod deletion, the device might be deleted while it is attached to a node,
  - If the pvc is used in another pod scheduled on a different node before the publish-failed pod deletion, the device might be attached to multiple nodes. 

@vladimirvivien

Is my understanding correct? And are there any issues that you can think of by not staging the volume there?

mkimuram commented:
@rootfs

> I see this PR does two things: a csi block refactoring and the operator generator reordering. What if breaking them up in separate PRs?

Thank you for your suggestion.
It makes sense to separate descriptor lock issue to another PR, for drivers that can detach the device while descriptor lock is held won't be affected by this issue. I will open another issue for descriptor lock and remove the code for this issue in this PR.

Even by separating descriptor lock issue,  fix for below operation_generator.go code would remain. However, the change become minimal and it would be more obvious to safe to change so than this issue.
> (a) Whether first arugument for NewBlockVolumeUnmapper in GenerateUnmapDeviceFunc can be changed safely,

mkimuram commented:
@rootfs @vladimirvivien 

I opened a separate issue for descriptor lock  #69114 and revert the change for it in this PR.
PTAL

mkimuram commented:
@rootfs @vladimirvivien 

Now, descriptor lock release issue was resolved in another PR, we can focus on rest of the issues and continue discussion here.  

What will be the next step to make a progress for this PR?

vladimirvivien commented:
@mkimuram apologies for the delay in response. Catching up with your PR.

vladimirvivien commented on pkg/volume/csi/csi_block.go:
```diff
@@ -55,41 +59,35 @@ func (m *csiBlockMapper) GetGlobalMapPath(spec *volume.Spec) (string, error) {
+	return path.Join(m.plugin.host.GetVolumeDevicePluginDir(csiPluginName), "staging", sanitizedSpecVolID)
+}
+
+// getPublishPath returns a path (on the node) to a device file which will be symlinked to
```
Please clarify the comment for func `getStagingPath` and `getPublishPath`.   They have the same comment.

vladimirvivien commented on pkg/volume/csi/csi_block.go:
```diff
@@ -55,41 +59,35 @@ func (m *csiBlockMapper) GetGlobalMapPath(spec *volume.Spec) (string, error) {
-		glog.Error(log("blockMapper.SetupDevice failed to get CSI persistent source: %v", err))
-		return "", err
-	}
+func (m *csiBlockMapper) stageVolumeForBlock(csiSource *v1.CSIPersistentVolumeSource, attachment *storage.VolumeAttachment) (string, error) {
```
Add short comment describing stageVolumeForBlock

vladimirvivien commented on pkg/volume/csi/csi_block.go:
```diff
@@ -132,145 +116,82 @@ func (m *csiBlockMapper) SetUpDevice() (string, error) {
-		glog.Error(log("blockMapper.MapDevice failed to get CSI persistent source: %v", err))
-		return err
-	}
+func (m *csiBlockMapper) publishVolumeForBlock(csiSource *v1.CSIPersistentVolumeSource, attachment *storage.VolumeAttachment, stagingPath string) (string, error) {
```
Add short comment describing func publishVolumeForBlock

vladimirvivien commented on pkg/volume/csi/csi_block.go:
```diff
@@ -279,46 +200,121 @@ func (m *csiBlockMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, vol
+		glog.Error(log("blockMapper.SetupDevice unable to find VolumeAttachment [id=%s]", attachID))
+		return "", errors.New("no existing VolumeAttachment found")
+	}
+
```
Since calls to `m.stageVolumeForBlock` and `m.publishVolumeForBlock` happen successively at close proximity, I would recommend to create `csiClient` here (instead of creating an instance in each method).  Then pass the client as a parameter to these methods respectively.

vladimirvivien commented on pkg/volume/csi/csi_block.go:
```diff
@@ -279,46 +200,121 @@ func (m *csiBlockMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, vol
 
-	glog.V(4).Infof(log("blockMapper.TearDownDevice NodeUnpublished successfully [%s]", podVolumeMapPath))
+	glog.V(4).Infof(log("unmapper.TearDownDevice(globalMapPath=%s; devicePath=%s)", globalMapPath, devicePath))
+
```
Same here. Since calls for `unpublishVolumeForBlock` and `unstageVolumeForBlock` happen successively, create a `csiClient` here and pass it to each method as a parameter respectively to avoid creating client inside each call.

vladimirvivien commented on pkg/volume/csi/csi_block.go:
```diff
@@ -279,46 +200,121 @@ func (m *csiBlockMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, vol
+
+	// Call NodeUnpublishVolume
+	publishPath := m.getPublishPath()
+	if _, err := os.Stat(publishPath); err == nil {
```
This `if` statement is swallowing the error when `err != nil`. So, rearrange this block so that it handles cases when `err != nil`.

vladimirvivien commented on pkg/volume/csi/csi_block.go:
```diff
@@ -279,46 +200,121 @@ func (m *csiBlockMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, vol
+
+	// Call NodeUnstageVolume
+	stagingPath := m.getStagingPath()
+	if _, err := os.Stat(stagingPath); err == nil {
```
Same here, rearrange the `if` block to handle cases when `err != nil`

vladimirvivien reviewed (commented):
This is getting close.  Thank you for your effort.  I have left a few suggestions and fixes.

mkimuram commented:
@vladimirvivien

Thank you for your review and detailed comment. I will fix as suggested.

mkimuram commented:
@vladimirvivien 

I fixed as suggested. PTAL

Also, I recently found an old discussion which seems to related to the implementation of the block volume support in k8s csi driver. In this thread, it only discuss on bind mount, not symlink.

https://github.com/container-storage-interface/spec/issues/80

So, I started to think that we need to decide whether we should allow both bind mount and symlink or only either of them. Also, if we allow bind mount, we need to decide who should create a file in mount point.

(Current implementation should work well for both, if drivers which use bind mount will create mount point by themselves. If we create a file in k8s side, then drivers which use symlink will need to delete a file to create symlink there.)

vladimirvivien commented:
@mkimuram let me take a look and will get back to you re bindmount vs symlink.

vladimirvivien commented on pkg/volume/csi/csi_block.go:
```diff
@@ -299,26 +297,24 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
-		glog.Error(log("blockMapper.TearDownDevice failed: %v", err))
+	// Call NodeUnstageVolume
+	stagingPath := m.getStagingPath()
+	if _, err := os.Stat(stagingPath); err != nil {
```
Decouple this if-else blocks since they (os.Stat check and unstageVol call) are independent.  In addition the stat check should only fail if err is not IsNotExist(err)

Suggestion:

```go
if _, err := os.Stat(...); os.IsNotExist(err) {
    glog.V(4).Infof(log("blockMapper.TearDownDevice stagingPath(%s) has already been deleted..."))
} else{
    return err
}
if err := m.unstageVolumeForBlock(ctx, m.csiClient, stagingPath); err != nil {
    return err
}
```

mkimuram replied:
@vladimirvivien 

Thank you for your suggestion.

I agree that os.stat check should only fail if err is not IsNotExist(err).

For IsNotExist(err) case, I intended to skip calling unstageVolumeForBlock because main purpose for unstageVolumeForBlock is to make staging path invisible, but the path does not exist already.

So, I fixed it as so. And I also applied the same change to unpublishVolumeForBlock.

PTAL

vladimirvivien reviewed (commented):
This looks great.  Only one suggestion and it should be LGTM'd.

vladimirvivien commented:
@mkimuram thanks for the thread you sent regarding how plugin should handle mounting device (bind mount vs symlink).  https://github.com/container-storage-interface/spec/issues/80 seems to be inconclusive, but seemed to have been leaning toward symlink.

The current alpha version of CSI block assumes a symlink and k8s creates and sends a File path as the target path during NodeStageVolume https://github.com/kubernetes/kubernetes/blob/4d66eeb92221e7228058b08a14a7078291edef85/pkg/volume/csi/csi_block.go#L160

Therefore, I think we should make sure your changes use symlink to ensure existing block implementations are not broken.

This is great work.  Once your code is merged, it will be an integral part in moving the CSI block support forward.

vladimirvivien commented:
@mkimuram Let me know when this is ready for review

mkimuram commented:
@vladimirvivien 

Thank you for your reminder. I meant that it's ready to review already with https://github.com/kubernetes/kubernetes/pull/68635#discussion_r226510749 , but it looks hidden in another comment.

vladimirvivien commented:
@mkimuram Ok, I will take one last look and approve it.

vladimirvivien commented:
@mkimuram these changes are great.  However, removing the responsibility of creating the symlink file from this code and letting drivers do it, will break at least 1 driver that already has that functionality.  This coming week I will do the followings
- Have the existing block driver test this PR
- Bring up the issue of symlink delation to the driver instead of k8s creating it
I hope to get this resolved this week so this can go in.

vladimirvivien commented:
@mkimuram are you on Slack? If you are or able to, I would like to further discuss the simlink vs bind mount issue on the WG-CSI channel.

vladimirvivien commented:
@mkimuram Also, since these changes will impact all drivers (not just RBD), can you please rename the title of the PR to something more appropriate (i.e. ` CSI block volume refactor to fix target path`)

vladimirvivien commented:
@mkimuram would you be able to join a CSI meeting this week to discuss your changes ? Let me know.  You can find me on Slack @vladv

mkimuram commented:
Today's  CSI meeting concluded that
  (1) Bind mount should be used instead of symlink,
  (2) CO need to prepare empty file in target path.

So, I will update this PR to reflect (2) and update rbd implementation to reflect (1) and confirm that it works well. In addition, I will also check that above change won't affect block volume code for existing in-tree driver.

mkimuram commented:
Currently, testing with below combination. It is working fine except that the pv remains un-deleted even after pvc deletion. I will continue investigation tomorrow.

- [k8s](https://github.com/kubernetes/kubernetes/pull/68635/commits/7a09199df4500f2fb804642f3c56e1f3b110dfc7)
- [csi rbd driver](https://github.com/mkimuram/ceph-csi/tree/blockv3)
- [csi provisioner](https://github.com/mkimuram/external-provisioner/commit/e94d21167fffef3270bca4c497b35aaf764c4ab2)

mkimuram commented:
@vladimirvivien @msau42 @lpabon @wongma7 @childsb @saad-ali 

Thank you for taking your time for discussion today.

Please also see below related discussions:
- CSI Spec: https://github.com/container-storage-interface/spec/issues/285
- Provisioner: https://github.com/kubernetes-csi/external-provisioner/pull/135

wongma7 commented:
This PR won't affect in-tree block volumes or moving the block volume feature to beta. My understanding: CSI drivers will create a file "plugins/kubernetes.io/csi/volumeDevices/publish/{volumeID}," and whether the file is a symlink or bind mount doesn't matter to operation_generator, because kubernetes will simply create symlinks to it at globalMapPath and volumeMapPath in MapDevice https://github.com/kubernetes/kubernetes/pull/68635/files#diff-ed824558531f8d1f616bc2f976266b64R286 like it currently does for every other in-tree plugin.

@msau42 So ref counting (& everything else) in operation_generator at https://github.com/kubernetes/kubernetes/blob/0f17e9ade6e089eadb7c09052a4a030f155a0eb0/pkg/volume/util/operationexecutor/operation_generator.go#L1073 will continue to work since there will continue to be symlinks at globalMapPath for that purpose

RE: taking lessons from this move from symlinks->bind mounting and applying that throughout the block volume code so that we can e.g. drop the loopback device code...IMO we can postpone discussion about that past 1.13 beta...because IMO while it may be a big change to make during a beta period, it's still an implementation detail from the user perspective & shouldn't block beta

msau42 commented:
Thanks for the analysis and testing @mkimuram and @wongma7!  Given this, I think it is fine to move in-tree block to beta in 1.13.  Something to keep in mind if we wanted to refactor this whole thing later, it would likely not support in-place upgrade scenarios (where node is not drained before updating kubelet).  According to @yujuhong, OSS does not officially support in-place upgrades, but people still do it, so we would need to make sure we at least keep in teardown code of the old methods so we don't leak devices/symlinks.  In-place downgrades would be extremely hard to cleanup automatically though...

mkimuram commented:
/retest

mkimuram commented:
@vladimirvivien @msau42 @lpabon @wongma7 @childsb @saad-ali

Sharing test results([csi-block-volume-test-log.txt](https://github.com/kubernetes/kubernetes/files/2554534/csi-block-volume-test-log.txt)) with below codes:

- [k8s](https://github.com/kubernetes/kubernetes/pull/68635/commits/7a09199df4500f2fb804642f3c56e1f3b110dfc7) (This PR)
- [csi rbd driver](https://github.com/ceph/ceph-csi/pull/97/commits/ff433f7fc2ecb5f673d7dba2f2c851cea6d4e4e1) (PR: https://github.com/ceph/ceph-csi/pull/97)
- [csi provisioner](https://github.com/mkimuram/external-provisioner/commit/e94d21167fffef3270bca4c497b35aaf764c4ab2) (PR: https://github.com/kubernetes-csi/external-provisioner/pull/135)

As seen in the test results, block volume is provisioned, mapped to the pods, and then deleted on pvc deletion, properly. Also, csi driver uses bind mount for staging/publish. Therefore, I think that this PR and provisioner PR are ready to merge. PTAL

bswartz commented:
I've tested this too on my setup and it works fine. I'm using my own test scripts however -- what are you using mkimuram?

vladimirvivien commented:
/lgtm 
/approve

bswartz commented:
/lgtm

AishSundar commented:
/milestone v1.13

AishSundar commented:
/priority important-soon

vladimirvivien commented:
@mkimuram please refactor so @saad-ali can approve.  Thanks.

mkimuram commented:
@bswartz @vladimirvivien 

Rebased and squashed the commit. PTAL
The conflicted part was just [s/glog/klog/ change](https://github.com/kubernetes/kubernetes/commit/954996e231074dc7429f7be1256a579bedd8344c#diff-ed824558531f8d1f616bc2f976266b64R54), so there's no logic change from the previous commit.

bswartz commented:
Tested again. Still good.

/lgtm

bswartz reviewed (changes requested):
In the 11/14 CSI meeting we agreed that staging paths should always be directories, even for blocks. This PR clarifies the wording: https://github.com/container-storage-interface/spec/pull/335
To conform to the spec this PR should be updated to create a directory for the staging path.

mkimuram committed:
Change stagingPath to be a directory

mkimuram commented:
@bswartz 

Changed ```stagingPath``` to a directory and also changed the comments. PTAL

bswartz commented:
@mkimuram We finally settled on another spec change which is that the CO should not create the 0-byte file during NodePublishVolume() either. It will be the SP's responsibility to create and delete the file at target_path. The CO must create the parent directory of target_path.

mkimuram committed:
Delegate creation of publishPath to driver and allow creating device file directly

mkimuram commented:
@bswartz 

> We finally settled on another spec change which is that the CO should not create the 0-byte file during NodePublishVolume() either. It will be the SP's responsibility to create and delete the file at target_path. The CO must create the parent directory of target_path.

Updated. PTAL

- [k8s](https://github.com/kubernetes/kubernetes/pull/68635/commits/6d7b5f1d1b104931072a006b1d986b8bd14f27d0) (This PR) ... updated
- [csi rbd driver](https://github.com/ceph/ceph-csi/pull/97/commits/e3a1d9fb40de1ce243bb18fb315d44fc16a2c0e1) (PR: https://github.com/ceph/ceph-csi/pull/97) .. updated
- [csi provisioner](https://github.com/mkimuram/external-provisioner/commit/e94d21167fffef3270bca4c497b35aaf764c4ab2) (PR: https://github.com/kubernetes-csi/external-provisioner/pull/135)

I've updated the csi rbd driver to create/delete target path, then PV deletion starts to fail again. 
However, this is driver side issue, not k8s side issue, so I'm sharing the current version.

I will look into the driver issue, tomorrow.

bswartz commented:
I updated my test driver to match the spec, and ran it against this patch, and it worked great. That's all I have time for tonight. If there's time tomorrow I'll take a look at the code changes.

bswartz commented:
/lgtm

saad-ali commented:
/approve

saad-ali commented:
/approve cancel

> target_path for publish: plugins/kubernetes.io/csi/volumeDevices/publish/{volumeID}

`staging_target_path` can be volume specific bug `target_path` for publish has to be pod specific

saad-ali commented on pkg/volume/csi/csi_block.go:
```diff
@@ -47,80 +51,61 @@ type csiBlockMapper struct {
+}
+
+// getPublishPath returns a publish path for a file (on the node) that should be used on NodePublishVolume/NodeUnpublishVolume
+// Example: plugins/kubernetes.io/csi/volumeDevices/publish/{volumeID}
```
If multiple containers use this, it will collide.

mkimuram replied:
@saad-ali 

Thank you for your comment.

Is my understanding right that you are suggesting that we should use ```spec.PersistentVolume.Name``` instead of ```kstrings.EscapeQualifiedNameForDisk(m.specName)```, as it is done in ```makeDeviceMountPath``` for filesystem volume?

Then, it will be like as below:
- block: ```/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/{pvname}```
             ```/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}```
- filesystem: ```/var/lib/kubelet/plugins/kubernetes.io/csi/pv/{pvname}/globalmount```


```m.specName``` is originally used there, because in-tree block volume code use it for ```GlobalMapPath``` and ```PodDeviceMapPath```.
- [GlobalMapPath](https://github.com/kubernetes/kubernetes/pull/68635/files#diff-ed824558531f8d1f616bc2f976266b64R54) : ```/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/{volumeID}/dev```
- [PodDeviceMapPath](https://github.com/kubernetes/kubernetes/pull/68635/files#diff-ed824558531f8d1f616bc2f976266b64R77): ```/var/lib/kubelet/pods/{podUid}/volumeDevices/kubernetes.io~csi/{volumeID}```

However, as explained in https://github.com/kubernetes/kubernetes/pull/68635#issuecomment-423259818, for block volume, we choose not to let csi driver handle ```GlobalMapPath``` and ```PodDeviceMapPath``` directly, instead csi driver make block volume visible to another path (```target_path```), then k8s (```ioutil.MapBlockVolume```) make ```target_path visible``` from ```GlobalMapPath``` and ```PodDeviceMapPath```. So, there is no restriction for using ```m.specName``` for ```target_path``` and ```staging_target_path```.

@bswartz @vladimirvivien 
Any concern?

mkimuram replied:
@saad-ali 

I checked the log and code again and found out that block codes already use the same ID to filesystem codes. I mean ```{pvname}``` should be the same to ```{volumeID}```, like ```pvc-a618a282-e929-11e8-bb66-525400b854f0```, so it should not collide.

Is it ok to just update the comment to make them look same?

saad-ali replied:
Would `spec.PersistentVolume.Name` collide if the PVC is used by multiple pods? Ideally the `target_path` should be unique per pod (have the pod ID in it) right? Otherwise if 2 pods use the same volume on the same node, each pod will have it's volume setup code running in parallel. 

I'm basing this on my understanding of mount, so it may well be different for block. If so I'd like to understand why?

mkimuram committed:
Fix the comment for csi block pathes to make look the same to filesystem

mkimuram commented:
@saad-ali @bswartz @vladimirvivien 

Fix the comment. PTAL

mkimuram commented:
/retest

AishSundar commented:
@saad-ali @bswartz @vladimirvivien is this fix good to go?

saad-ali commented:
/approve
